File: scratchpad/tiling4_20260912/hex_parity.py
"""TILING4 -- the prototype half of the hex-tiling parity check.

It evaluates h_cand.py's OWN `tri_grid` and `hash01` (imported, not re-typed --
if the prototype is edited this check follows it) plus the same blend formula,
at the same fifteen world positions x five settings as hex_parity.cpp, and diffs
the two.

FLOORS, asserted rather than eyeballed:
  * the OFF block must equal the plain wrapped tap exactly -- "off is the rung's
    bytes" is this switch's whole safety argument;
  * every ON block must DIFFER from OFF at a majority of the fifteen positions,
    or the comparison is passing on a sampler that does nothing;
  * a deliberately WRONG C++ (the skew's sign flipped) must fail this check --
    run with --sabotage to see it fail, which is the only proof the check can
    fail at all.

    python hex_parity.py            ->  logs/hex_parity.txt
    python hex_parity.py --sabotage ->  the same check against a wrong prototype
"""
import os
import subprocess
import sys

import numpy as np

HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, HERE)
sys.path.insert(0, os.path.join(os.path.dirname(HERE), 'tiling3_20260911'))
sys.path.insert(0, os.path.join(os.path.dirname(HERE), 'tiling2_20260911'))
sys.path.insert(0, os.path.join(os.path.dirname(HERE), 'splat1_20260911'))
import h_cand as H                                            # noqa: E402

TILE = 341.3333
POS = [(0.0, 0.0), (1.0, -1.0), (341.3333, 341.3333),
       (-81920.0, 98304.0), (-81920.0, 81920.0), (-147456.0, -81920.0),
       (-16384.0, -81920.0), (114688.0, -81920.0), (-16384.0, 65536.0),
       (98304.0, 65536.0), (1999999.0, -1999999.0), (12345.678, -98765.432),
       (-0.5, -0.5), (-1024.0, -1024.0), (1023.9999, 1023.9999)]
SIZES = [np.float32(256.0), np.float32(341.3333), np.float32(512.0),
         np.float32(682.6667)]
MEAN = np.array([127.5, 127.5, 127.5])


def fake_tap(u, v):
    return np.array([255.0 * u, 255.0 * v, 255.0 * u * v])


def wrap(x, tile=TILE):
    u = np.fmod(x / tile, 1.0)
    return u + 1.0 if u < 0.0 else u


def py_off(wx, wy):
    return fake_tap(wrap(wx), wrap(wy))


def py_hex(wx, wy, size, sabotage=False):
    """The prototype's own tri_grid and hash01, on one scalar position."""
    px = np.array([wx], np.float64) / np.float64(size)
    py = np.array([wy], np.float64) / np.float64(size)
    if sabotage:
        # the deliberate defect: the skew applied with the wrong sign
        sx = px + 0.57735026918962576 * py
        sy = 1.15470053837925152 * py
        bi = np.floor(sx)
        bj = np.floor(sy)
        tx, ty = sx - bi, sy - bj
        tz = 1.0 - tx - ty
        up = tz > 0.0
        w = (np.where(up, tz, -tz), np.where(up, ty, 1.0 - ty),
             np.where(up, tx, 1.0 - tx))
        o = np.where(up, 0.0, 1.0)
        verts = (((bi + o).astype(np.int64), (bj + o).astype(np.int64)),
                 ((bi + o).astype(np.int64), (bj + 1.0 - o).astype(np.int64)),
                 ((bi + 1.0 - o).astype(np.int64), (bj + o).astype(np.int64)))
    else:
        w, verts = H.tri_grid(px, py)
    acc = np.zeros(3)
    wsq = 0.0
    cells = []
    for k, (i, j) in enumerate(verts):
        ox = H.hash01(i, j, 0)[0] * TILE
        oy = H.hash01(i, j, 1)[0] * TILE
        s = fake_tap(wrap(wx + ox), wrap(wy + oy))
        acc += (s - MEAN) * float(w[k][0])
        wsq += float(w[k][0]) ** 2
        cells.append((int(i[0]), int(j[0])))
    if wsq > 1e-12:
        acc = acc / np.sqrt(wsq)
    return cells, [float(x[0]) for x in w], MEAN + acc


def main():
    sab = '--sabotage' in sys.argv
    cpp = os.path.join(HERE, 'logs', 'hex_parity_cpp.txt')
    if not os.path.exists(cpp):
        print('REFUSED: %s is missing -- compile and run hex_parity.cpp first'
              % cpp)
        return 1
    lines = [l.strip() for l in open(cpp) if l.strip()]
    L = ['TILING4 -- hex tiling: src/lodgen.cpp vs the prototype', '',
         'fifteen world positions x five settings; the C++ side is',
         'hex_parity.cpp (lodgenLandHexCell and lodgenLandHexOffset copied',
         'verbatim), the prototype side is h_cand.py`s own tri_grid and hash01.',
         'Both run in double: this checks the geometry, the hash, the vertex',
         'triple and the blend formula, not the product`s last float bit.', '']
    if sab:
        L.append('--sabotage: the prototype side has the SKEW SIGN FLIPPED, so')
        L.append('this run MUST fail -- it is the proof the check can fail.')
        L.append('')

    it = iter(lines)
    head = next(it)
    assert head == 'OFF', head
    bad = 0
    off = []
    for n, (wx, wy) in enumerate(POS):
        got = [float(x) for x in next(it).split()]
        want = py_off(wx, wy)
        off.append(got)
        d = max(abs(g - w) for g, w in zip(got, want))
        if d > 1e-6:
            bad += 1
            L.append('   OFF (%g,%g): C++ %s vs prototype %s' % (wx, wy, got, want))
    L.append('OFF: %d of 15 positions disagree (floor 0).' % bad)

    nmoved = []
    for s, size in enumerate(SIZES):
        head = next(it)
        assert head.startswith('SIZE='), head
        b = 0
        moved = 0
        for n, (wx, wy) in enumerate(POS):
            f = next(it).split()
            cvi = [(int(f[0]), int(f[1])), (int(f[2]), int(f[3])),
                   (int(f[4]), int(f[5]))]
            cw = [float(x) for x in f[6:9]]
            co = [float(x) for x in f[9:12]]
            pcells, pw, po = py_hex(wx, wy, size, sab)
            dcell = 0 if cvi == pcells else 1
            dw = max(abs(a - b2) for a, b2 in zip(cw, pw))
            do = max(abs(a - b2) for a, b2 in zip(co, po))
            if dcell or dw > 1e-9 or do > 1e-6:
                b += 1
                L.append('   size %g (%g,%g): cells %s vs %s, dw %.3g, dcol %.3g'
                         % (float(size), wx, wy, cvi, pcells, dw, do))
            if max(abs(a - b2) for a, b2 in zip(co, off[n])) > 1e-6:
                moved += 1
        nmoved.append(moved)
        L.append('size %-9g %d of 15 positions disagree (floor 0); the sampler '
                 'moved %d of 15 off the plain tap (floor 8)'
                 % (float(size), b, moved))
        bad += b

    L.append('')
    ok = bad == 0 and all(m >= 8 for m in nmoved)
    L.append('PARITY: %s -- %d disagreements over 75 comparisons, and every '
             'setting moves the sample.' % ('PASS' if ok else 'FAIL', bad))
    if sab:
        L.append('(and with the sabotage in place a PASS here would mean the '
                 'check cannot fail)')
    txt = '\n'.join(L) + '\n'
    print(txt)
    name = 'hex_parity_sabotage.txt' if sab else 'hex_parity.txt'
    with open(os.path.join(HERE, 'logs', name), 'w', newline='\n') as f:
        f.write(txt)
    return 0 if ok else 2


if __name__ == '__main__':
    sys.exit(main())

File: scratchpad/tiling4_20260912/hex_parity.cpp
/* TILING4 -- does src/lodgen.cpp's hex tiling compute the same numbers as the
 * prototype the numbers were picked on?
 *
 * The C++ that landed is a transcription of h_cand.py's `tri_grid`, `hash01`
 * and `_blend`.  A transcription is a place to make a mistake -- a skew the
 * wrong way round, a vertex triple in the wrong order, a weight taken from the
 * mirrored triangle, a missing square root -- and none of those would crash:
 * they would quietly bake a DIFFERENT sampler from the one the fourteen sheets
 * were measured on, and every number in this lane's report would then describe
 * a bake the exe never produces.
 *
 * So the same pure function is evaluated on both sides at the same fifteen
 * world positions x five settings and diffed.  lodgenLandHexCell and
 * lodgenLandHexOffset are copied VERBATIM from src/lodgen.cpp, with Qt's
 * typedefs spelled out.  The texture is replaced by a closed-form stand-in --
 * s(u,v) = (255u, 255v, 255uv) -- so the blend's arithmetic is exercised with
 * no DDS, no Qt and no link: pure IEEE double on both sides.
 *
 * WHAT THIS PROBE CANNOT REACH.  In the product the three taps come back as
 * float32 from getPixelT and the accumulator is a FloatVector4, so the shipped
 * blend runs in float; here both sides run in double.  This proves the
 * GEOMETRY, THE HASH, THE VERTEX TRIPLE AND THE BLEND FORMULA agree; it does
 * not claim the product's last float bit.  That is what gate F2 is for.
 *
 *   (compile)  g++ -O2 -std=c++17 -static -o hex_parity.exe hex_parity.cpp
 *   (run)      hex_parity.exe > logs/hex_parity_cpp.txt
 */
#include <cstdio>
#include <cmath>
#include <cstdint>
typedef int32_t qint32;
typedef uint32_t quint32;

static float g_landHexSize = 0.0f;

/* ---- verbatim from src/lodgen.cpp -------------------------------------- */
static inline quint32 lodgenWarpHash( qint32 i, qint32 j, quint32 k )
{
	quint32 h = quint32( i ) * 374761393u
		+ quint32( j ) * 668265263u
		+ k * 2246822519u;
	h ^= h >> 13;
	h *= 1274126177u;
	h ^= h >> 16;
	return h;
}

static const double LODGEN_HEX_SKEW  = 0.57735026918962576;
static const double LODGEN_HEX_SCALE = 1.15470053837925152;

static void lodgenLandHexCell( double wx, double wy, double size,
                               qint32 * vi, qint32 * vj, double * w )
{
	const double px = wx / size;
	const double py = wy / size;
	const double sx = px - LODGEN_HEX_SKEW * py;
	const double sy = LODGEN_HEX_SCALE * py;
	const double bi = std::floor( sx );
	const double bj = std::floor( sy );
	const double tx = sx - bi;
	const double ty = sy - bj;
	const double tz = 1.0 - tx - ty;
	const bool up = tz > 0.0;
	const double o = up ? 0.0 : 1.0;
	w[0] = up ? tz : -tz;
	w[1] = up ? ty : 1.0 - ty;
	w[2] = up ? tx : 1.0 - tx;
	vi[0] = qint32( bi + o );        vj[0] = qint32( bj + o );
	vi[1] = qint32( bi + o );        vj[1] = qint32( bj + 1.0 - o );
	vi[2] = qint32( bi + 1.0 - o );  vj[2] = qint32( bj + o );
}

static inline double lodgenLandHexOffset( qint32 i, qint32 j, quint32 k )
{
	return double( lodgenWarpHash( i, j, k ) ) / 4294967296.0;
}
/* ---- end verbatim ------------------------------------------------------ */

/* the stand-in for getPixelT: a closed form, so both languages agree exactly */
static void fakeTap( double u, double v, double * rgb )
{
	rgb[0] = 255.0 * u;
	rgb[1] = 255.0 * v;
	rgb[2] = 255.0 * u * v;
}

/* the shipped blend, with the stand-in in place of the texture */
static void hexSample( double wx, double wy, double tile, double * out )
{
	if ( g_landHexSize <= 0.0f ) {
		double u = std::fmod( wx / tile, 1.0 );
		double v = std::fmod( wy / tile, 1.0 );
		if ( u < 0.0 ) u += 1.0;
		if ( v < 0.0 ) v += 1.0;
		fakeTap( u, v, out );
		return;
	}
	qint32 vi[3], vj[3];
	double w[3];
	lodgenLandHexCell( wx, wy, double( g_landHexSize ), vi, vj, w );
	const double mean[3] = { 127.5, 127.5, 127.5 };
	double acc[3] = { 0.0, 0.0, 0.0 };
	double wsq = 0.0;
	for ( int k = 0; k < 3; k++ ) {
		const double ox = lodgenLandHexOffset( vi[k], vj[k], 0 ) * tile;
		const double oy = lodgenLandHexOffset( vi[k], vj[k], 1 ) * tile;
		double tu = std::fmod( ( wx + ox ) / tile, 1.0 );
		double tv = std::fmod( ( wy + oy ) / tile, 1.0 );
		if ( tu < 0.0 ) tu += 1.0;
		if ( tv < 0.0 ) tv += 1.0;
		double s[3];
		fakeTap( tu, tv, s );
		for ( int c = 0; c < 3; c++ )
			acc[c] += ( s[c] - mean[c] ) * w[k];
		wsq += w[k] * w[k];
	}
	if ( wsq > 1e-12 ) {
		const double inv = 1.0 / std::sqrt( wsq );
		for ( int c = 0; c < 3; c++ )
			acc[c] *= inv;
	}
	for ( int c = 0; c < 3; c++ )
		out[c] = mean[c] + acc[c];
}

int main()
{
	const double TILE = 341.3333;
	const double P[15][2] = {
		{ 0.0, 0.0 }, { 1.0, -1.0 }, { 341.3333, 341.3333 },
		{ -81920.0, 98304.0 }, { -81920.0, 81920.0 }, { -147456.0, -81920.0 },
		{ -16384.0, -81920.0 }, { 114688.0, -81920.0 }, { -16384.0, 65536.0 },
		{ 98304.0, 65536.0 }, { 1999999.0, -1999999.0 }, { 12345.678, -98765.432 },
		{ -0.5, -0.5 }, { -1024.0, -1024.0 }, { 1023.9999, 1023.9999 }
	};
	const float SIZES[4] = { 256.0f, 341.3333f, 512.0f, 682.6667f };

	/* OFF first: it must be the plain wrapped tap, which is what the shipped
	 * lodgenLandHexTap returns by an early `return` rather than by arithmetic */
	printf( "OFF\n" );
	for ( int n = 0; n < 15; n++ ) {
		double o[3];
		hexSample( P[n][0], P[n][1], TILE, o );
		printf( "%.9g %.9g %.9g\n", o[0], o[1], o[2] );
	}
	for ( int s = 0; s < 4; s++ ) {
		g_landHexSize = SIZES[s];
		printf( "SIZE=%.9g\n", double( SIZES[s] ) );
		for ( int n = 0; n < 15; n++ ) {
			/* the lattice cell and the weights, then the blended colour: a
			 * wrong vertex triple with right weights would otherwise hide */
			qint32 vi[3], vj[3];
			double w[3], o[3];
			lodgenLandHexCell( P[n][0], P[n][1], double( g_landHexSize ),
			                   vi, vj, w );
			hexSample( P[n][0], P[n][1], TILE, o );
			printf( "%d %d %d %d %d %d %.9g %.9g %.9g %.9g %.9g %.9g\n",
			        int( vi[0] ), int( vj[0] ), int( vi[1] ), int( vj[1] ),
			        int( vi[2] ), int( vj[2] ), w[0], w[1], w[2],
			        o[0], o[1], o[2] );
		}
	}
	return 0;
}
